index/secure_index.go

```diff
@@ -0,0 +1,64 @@
+package index
+
+import (
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/binary"
+	"fmt"
+	"github.com/jxguan/go-datastructures/bitarray"
+	"hash"
+)
+
+// SecureIndex defines the elements in a secure index.
+type SecureIndex struct {
+	BloomFilter bitarray.BitArray // The blinded bloom filter, which is the main part of the index.
+	DocID       int               // The document ID that this index is for.
+	Size        uint64            // The number of buckets in the bloom filter.
+	Hash        func() hash.Hash  // The hash function to be used for HMAC.
+}
+
+// Marshal serializes a SecureIndex into a byte slice.
+func (si *SecureIndex) Marshal() []byte {
```

Comment:
Implementing the [`encoding.BinaryMarshaler`](https://golang.org/pkg/encoding/#BinaryMarshaler) interface would be more standard.

(Same with Unmarshal below.)